src/routes/index.tsx

```diff
@@ -376,6 +376,16 @@ class PageErrorBoundary extends Component<
 
   componentDidCatch(error: Error | null, info: object) {
     this.setState({ error, info })
+
+    if (!window.requestAnimationFrame) {
+      setTimeout(() => {
+        this.setState({ error: undefined, info: { componentStack: '' } })
+      }, 100)
+    } else {
+      requestAnimationFrame(() => {
+        this.setState({ error: undefined, info: { componentStack: '' } })
+      })
+    }
```

Comment:
In theory, when error is set to null, the PageErrorBoundary will re-render, but due to the condition if (!error) return children, it will render the old component again, triggering another rendering error and returning to the PageErrorBoundary, creating an infinite loop.

I tested it locally, and it aligns with this theoretical expectation. However, in the video recording you mentioned, I did not observe this specific scenario. I am not sure of the exact reason for this discrepancy; it could be a special case. We should find a more reasonable approach to resolve this issue.

One possible solution is to listen for route changes within the PageErrorBoundary and automatically set error to null when a change occurs. Another approach could involve setting up a beforeRoute in the route configuration to clean up the data in the PageErrorBoundary. I would recommend the first approach because it avoids coupling between components.

Reply:
I have two possible solutions too:
1. Break the global `ErrorBoundary` into small ones, and keep `header` and `footer` independent.
2. Add a `homepage` link with `refresh` on the error page

Reply:
> 1. Break the global ErrorBoundary into small ones, and keep header and footer independent.

I'm not entirely sure if I understand the meaning of this solution, but currently, the `PageErrorBoundary` wraps around the `Route`.
If we don't change the hierarchy where the `PageErrorBoundary` is located, simply splitting the implementation of `return <NotFoundPage errorMessage={error.toString()} errorDescription={info.componentStack} />` will still result in the issue of invalid links when clicked. This is because the `error` in the `PageErrorBoundary` is not reset.

> 2. Add a homepage link with refresh on the error page

This is a concise implementation solution, but users may still be confused when other links become invalid.

Reply:
> > 1. Break the global ErrorBoundary into small ones, and keep header and footer independent.
> 
> I'm not entirely sure if I understand the meaning of this solution, but currently, the `PageErrorBoundary` wraps around the `Route`. If we don't change the hierarchy where the `PageErrorBoundary` is located, simply splitting the implementation of `return <NotFoundPage errorMessage={error.toString()} errorDescription={info.componentStack} />` will still result in the issue of invalid links when clicked. This is because the `error` in the `PageErrorBoundary` is not reset.
> 
> > 2. Add a homepage link with refresh on the error page
> 
> This is a concise implementation solution, but users may still be confused when other links become invalid.

For point 1, there will be 3 `ErrorBoundary` and the case will be

```
<ErrorBoundary>
  <Header />
</ErrorBoundary>

<ErrorBoundary>
  <ErrorMessage />
</ErrorBoundary>

<ErrorBoundary>
  <Footer />
</ErrorBoundary>
```

Only links in the middle part don't work, but there will be no links if error message is displayed.

For point 2, my bad for not covering other invalid links, it's not a good solution.

Reply:
> For point 1, there will be 3 ErrorBoundary and the case will be

This would require changing the hierarchy and design of the `ErrorBoundary`, such as how to handle header crashes. 
This could involve a significantly larger amount of work compared to resetting the `error` when the route changes, and the potential benefits may not outweigh the effort involved.

Reply:
I think it would be possible to add a new state attribute called "shouldReset" to indicate whether the error state should be reset. In the componentDidCatch method, we can set shouldReset to true, indicating that the error state needs to be reset. Then, in the componentDidUpdate method, we can check the value of shouldReset. If it's true, we can perform the reset operation by setting the error state and related information back to their initial values and setting shouldReset to false.
```ts
class PageErrorBoundary extends Component<
  {},
  {
    error?: Error | null;
    info: {
      componentStack?: string;
    };
    shouldReset: boolean;
  }
> {
  constructor(props: {}) {
    super(props);

    this.state = {
      error: null,
      info: {
        componentStack: '',
      },
      shouldReset: false,
    };
  }

  componentDidCatch(error: Error | null, info: object) {
    this.setState({ error, info, shouldReset: true });
  }

  componentDidUpdate() {
    const { shouldReset } = this.state;
    if (shouldReset) {
      this.setState({ error: null, info: { componentStack: '' }, shouldReset: false });
    }
  }

  render() {
    const { children } = this.props;
    const { error, info } = this.state;
    if (!error) return children;

    return <NotFoundPage errorMessage={error.toString()} errorDescription={info.componentStack} />;
  }
}
```

Reply:
2. By listening to the componentDidUpdate method and comparing prevProps with this.props, we can detect if an update has occurred. If an update is detected, the error and info states will be reset to their initial values, allowing for re-capturing and handling of errors the next time they occur.
```ts
type PageErrorBoundaryState = {
  error?: Error | null
  info: {
    componentStack?: string
  }
}

type PageErrorBoundaryProps = {}

class PageErrorBoundary extends Component<PageErrorBoundaryProps, PageErrorBoundaryState> {
  constructor(props: PageErrorBoundaryProps) {
    super(props)

    this.state = {
      error: null,
      info: {
        componentStack: '',
      },
    }
  }

  static getDerivedStateFromError(error: Error) {
    return { error }
  }

  componentDidUpdate(prevProps: PageErrorBoundaryProps) {
    if (prevProps !== this.props) {
      this.setState({ error: null, info: { componentStack: '' } })
    }
  }

  componentDidCatch(error: Error | null, info: React.ErrorInfo) {
    this.setState({ error, info })
  }

  render() {
    const { error, info } = this.state
    const { children } = this.props

    if (error) {
      return <NotFoundPage errorMessage={error.toString()} errorDescription={info.componentStack} />
    }

    return children
  }
}
```

Reply:
Regarding the solution of resetting the error in componentDidUpdate when the `children prop change`, I agree that it is a quick implementation.
However, I have concerns about potential unintended resets, as our actual requirement is to reset the error only when the user clicks a link or attempts to leave the error page. This requirement may not be entirely equivalent to the event of `children prop change`.

Reply:
> Regarding the solution of resetting the error in componentDidUpdate when the `children prop change`, I agree that it is a quick implementation. However, I have concerns about potential unintended resets, as our actual requirement is to reset the error only when the user clicks a link or attempts to leave the error page. This requirement may not be entirely equivalent to the event of `children prop change`.

Adding a homepage link and a refresh option on the error page might be suitable, but I believe the best solution for the user is to reset the error in componentDidUpdate. For users, being able to navigate is considered normal

Reply:
It looks like the discussion has already touched upon product interaction, any thoughts from @Danie0918 and @Sven-TBD

Reply:
Considering the above situation, can we hide the header and footer on the 404 page and provide a button to return to the home page.

![page3](https://github.com/Magickbase/ckb-explorer-frontend/assets/122437870/3b14bedf-7489-4bf5-896f-c350a7d2184e)

Reply:
Since it is an exception page, it is special to the user and it is understandable that there is no header and footer. What the user needs at this point is more help with resolution, so showing the cause of the error and providing suggested actions is what is needed most.

![image](https://github.com/Magickbase/ckb-explorer-frontend/assets/122437870/855d331f-e2b4-407b-a3c6-dd7f07ff47eb)

Reply:
> Considering the above situation, can we hide the header and footer on the 404 page and provide a button to return to the home page.
> 
> ![page3](https://user-images.githubusercontent.com/122437870/245048439-e1dbee45-aa2e-4d9c-9468-1a4db3d92297.png)

@Kirl70 Here you need to provide a UI diagram of an error page that does not contain headers and footers.

Reply:
Design Draft:
https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?type=design&node-id=17517%3A39525&t=TgkC3Pg1ICQwcqRt-1
<img width="1425" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/106060898/fbd7a1d3-7e63-49ee-9f73-92e4df82a7bf">